File: scripts/publish_family_update.py
#!/usr/bin/env python3
"""Trusted, hub-only PR publication. Never executes candidate component code."""
import argparse
import base64
import hashlib
import json
import os
from pathlib import Path
import re
import subprocess
import sys
import tomllib

from family import Family, FamilyError
from family_update import successful

REPO = "neverhuman/jankurai"
FILES = {"family.lock", "Cargo.lock"}


def api(endpoint, payload=None, method=None):
    args = ["gh", "api", endpoint]
    if method:
        args += ["--method", method]
    if payload is not None:
        args += ["--input", "-"]
    result = subprocess.run(args, input=json.dumps(payload) if payload is not None else None,
                            text=True, capture_output=True, check=False)
    if result.returncode:
        raise FamilyError(result.stderr.strip())
    return json.loads(result.stdout) if result.stdout.strip() else None


def content(path, ref):
    data = api(f"repos/{REPO}/contents/{path}?ref={ref}")
    return base64.b64decode(data["content"]).decode()


def validate_candidate(family, candidate, base):
    old, new = tomllib.loads(base["family.lock"]), tomllib.loads(candidate["family.lock"])
    if set(old) != set(new) or any(old[k] != new[k] for k in old if k != "repo"):
        raise FamilyError("updater may only change component tag/commit pins")
    if len(new["repo"]) != len(old["repo"]):
        raise FamilyError("candidate changed family membership")
    changed = False
    for previous, pin in zip(old["repo"], new["repo"]):
        if set(previous) != set(pin) or any(previous[k] != pin[k] for k in previous if k not in ("tag", "commit")):
            raise FamilyError("candidate changed repository metadata/order")
        if pin == previous:
            continue
        changed = True
        sha = pin["commit"]
        if not re.fullmatch(r"[0-9a-f]{40}", sha) or pin["tag"] != f"ci-{sha}":
            raise FamilyError("candidate needs an immutable exact-SHA CI tag")
        repo = next(r for r in family.repos if r["name"] == pin["repo"])
        ref = api(f"repos/{repo['slug']}/git/ref/tags/{pin['tag']}")
        if ref["object"]["type"] != "commit" or ref["object"]["sha"] != sha:
            raise FamilyError("candidate CI tag mismatch")
        comparison = api(f"repos/{repo['slug']}/compare/{sha}...{repo['default_branch']}")
        if comparison["status"] not in ("ahead", "identical") or not successful(repo, sha):
            raise FamilyError("candidate is not a successful default-branch revision")
    if not changed:
        raise FamilyError("candidate contains no component update")
    cargo = tomllib.loads(candidate["Cargo.lock"])
    if cargo.get("version") != 4 or not isinstance(cargo.get("package"), list):
        raise FamilyError("invalid aggregate Cargo lock")


def branch_for(candidate):
    digest = hashlib.sha256((candidate["family.lock"] + "\0" + candidate["Cargo.lock"]).encode()).hexdigest()
    return "automation/family-" + digest[:24]


def publish(family, directory):
    candidate = {}
    for name in FILES:
        path = directory / name
        if path.is_symlink() or not path.is_file() or path.stat().st_size > 2_000_000:
            raise FamilyError("invalid candidate artifact")
        candidate[name] = path.read_text()
    base_sha = api(f"repos/{REPO}/git/ref/heads/main")["object"]["sha"]
    base = {name: content(name, base_sha) for name in FILES}
    validate_candidate(family, candidate, base)
    branch = branch_for(candidate)
    existing = api(f"repos/{REPO}/pulls?state=open&head=neverhuman:{branch}")
    if existing:
        print(existing[0]["html_url"])
        return
    tree = []
    for name in sorted(FILES):
        blob = api(f"repos/{REPO}/git/blobs", {"content": candidate[name], "encoding": "utf-8"})
        tree.append({"path": name, "mode": "100644", "type": "blob", "sha": blob["sha"]})
    base_tree = api(f"repos/{REPO}/git/commits/{base_sha}")["tree"]["sha"]
    tree_sha = api(f"repos/{REPO}/git/trees", {"base_tree": base_tree, "tree": tree})["sha"]
    commit = api(f"repos/{REPO}/git/commits", {
        "message": "Update validated Jankurai family locks", "tree": tree_sha, "parents": [base_sha]})
    api(f"repos/{REPO}/git/refs", {"ref": f"refs/heads/{branch}", "sha": commit["sha"]})
    pr = api(f"repos/{REPO}/pulls", {
        "title": "Update validated Jankurai component revisions", "head": branch, "base": "main",
        "body": "Automated family update. Each changed component has a successful required check and "
                "an immutable CI tag. The candidate locks passed combined integration in a disposable "
                "CI checkout. Ordinary PR CI must pass before a protected merge."})
    print(pr["html_url"])


def merge(family):
    actor = api("user")["login"]
    hub = next(r for r in family.repos if r["name"] == "jankurai")
    for pr in api(f"repos/{REPO}/pulls?state=open&base=main&per_page=100"):
        if (pr["user"]["login"] != actor or pr["head"]["repo"]["full_name"] != REPO
                or not pr["head"]["ref"].startswith("automation/family-") or pr["draft"]):
            continue
        files = api(f"repos/{REPO}/pulls/{pr['number']}/files?per_page=100")
        if not files or any(f["filename"] not in FILES or f["status"] != "modified" for f in files):
            continue
        candidate = {name: content(name, pr["head"]["sha"]) for name in FILES}
        base = {name: content(name, pr["base"]["sha"]) for name in FILES}
        if branch_for(candidate) != pr["head"]["ref"]:
            continue
        validate_candidate(family, candidate, base)
        if not successful(hub, pr["head"]["sha"]):
            continue
        # GitHub enforces strict protected-branch checks; sha prevents a head race.
        result = api(f"repos/{REPO}/pulls/{pr['number']}/merge", {
            "sha": pr["head"]["sha"], "merge_method": "squash"}, method="PUT")
        if not result.get("merged"):
            raise FamilyError("protected updater merge was refused")
        print(f"Merged {pr['html_url']} at {result['sha']}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("command", choices=["publish", "merge"])
    parser.add_argument("--directory", type=Path, default=Path("target/candidate"))
    args = parser.parse_args()
    try:
        family = Family(Path(__file__).resolve().parent.parent)
        if args.command == "publish":
            publish(family, args.directory)
        else:
            merge(family)
    except (FamilyError, KeyError, ValueError, OSError) as exc:
        sys.exit(f"family automation: {exc}")

File: scripts/publish-family-update.mjs
import fs from 'node:fs';
import path from 'node:path';
import { createHash } from 'node:crypto';
import { fileURLToPath } from 'node:url';
import TOML from '@iarna/toml';
import { Family } from './family-model.mjs';
import { api, successful } from './family-update.mjs';
import { isLink } from './family-lib.mjs';

const repository = 'neverhuman/jankurai', files = ['family.lock', 'Cargo.lock'];
const content = (file, ref, request = api) => Buffer.from(request(`repos/${repository}/contents/${file}?ref=${ref}`).content, 'base64').toString();
export const branchFor = candidate => 'automation/family-' + createHash('sha256').update(candidate['family.lock'] + '\0' + candidate['Cargo.lock']).digest('hex').slice(0, 24);
const same = (left, right) => JSON.stringify(left) === JSON.stringify(right);
export function validateCandidate(family, candidate, base, request = api) {
  const old = TOML.parse(base['family.lock']), next = TOML.parse(candidate['family.lock']);
  if (!same(Object.keys(old), Object.keys(next)) || Object.keys(old).some(key => key !== 'repo' && !same(old[key], next[key]))) throw new Error('updater may only change component pins');
  if (old.repo.length !== next.repo.length) throw new Error('candidate changed family membership');
  let changed = false;
  for (let i = 0; i < old.repo.length; i++) {
    const previous = old.repo[i], pin = next.repo[i];
    if (!same(Object.keys(previous), Object.keys(pin)) || Object.keys(previous).some(key => !['tag', 'commit'].includes(key) && !same(previous[key], pin[key]))) throw new Error('candidate changed repository metadata/order');
    if (same(previous, pin)) continue;
    changed = true;
    validatePin(family, pin, request);
  }
  if (!changed) throw new Error('candidate contains no component update');
  const cargo = TOML.parse(candidate['Cargo.lock']);
  if (cargo.version !== 4 || !Array.isArray(cargo.package)) throw new Error('invalid aggregate Cargo lock');
}
function validatePin(family, pin, request) {
  const sha = pin.commit;
  if (!/^[a-f0-9]{40}$/.test(sha) || pin.tag !== `ci-${sha}`) throw new Error('candidate needs an immutable exact-SHA CI tag');
  const repo = family.repos.find(repo => repo.name === pin.repo);
  const ref = request(`repos/${repo.slug}/git/ref/tags/${pin.tag}`).object;
  if (ref.type !== 'commit' || ref.sha !== sha) throw new Error('candidate CI tag mismatch');
  const comparison = request(`repos/${repo.slug}/compare/${sha}...${repo.default_branch}`);
  if (!['ahead', 'identical'].includes(comparison.status) || !successful(repo, sha, request)) throw new Error('candidate is not a successful default-branch revision');
}
export function publish(family, directory, request = api) {
  const candidate = {};
  for (const file of files) {
    const name = path.join(directory, file);
    if (isLink(name) || !fs.statSync(name).isFile() || fs.statSync(name).size > 2_000_000) throw new Error('invalid candidate artifact');
    candidate[file] = fs.readFileSync(name, 'utf8');
  }
  const baseSha = request(`repos/${repository}/git/ref/heads/main`).object.sha;
  validateCandidate(family, candidate, Object.fromEntries(files.map(file => [file, content(file, baseSha, request)])), request);
  const branch = branchFor(candidate);
  const existing = request(`repos/${repository}/pulls?state=open&head=neverhuman:${branch}`);
  let previousHead;
  if (existing.length) {
    const pr = existing[0];
    const changed = request(`repos/${repository}/pulls/${pr.number}/files?per_page=100`);
    if (pr.head.repo?.full_name !== repository || pr.head.ref !== branch ||
        changed.some(file => !files.includes(file.filename) || file.status !== 'modified')) {
      throw new Error('existing updater PR contains unexpected changes');
    }
    for (const file of files) {
      if (content(file, pr.head.sha, request) !== candidate[file]) throw new Error('existing updater candidate changed');
    }
    const comparison = request(`repos/${repository}/compare/${baseSha}...${pr.head.sha}`);
    if (comparison.merge_base_commit.sha === baseSha) { console.log(pr.html_url); return; }
    previousHead = pr.head.sha;
  }
  const tree = files.map(file => ({ path: file, mode: '100644', type: 'blob', sha:
    request(`repos/${repository}/git/blobs`, { content: candidate[file], encoding: 'utf-8' }).sha }));
  const baseTree = request(`repos/${repository}/git/commits/${baseSha}`).tree.sha;
  const treeSha = request(`repos/${repository}/git/trees`, { base_tree: baseTree, tree }).sha;
  const commit = request(`repos/${repository}/git/commits`, { message: 'Update validated Jankurai family locks', tree: treeSha,
    parents: previousHead ? [previousHead, baseSha] : [baseSha] });
  if (previousHead) {
    request(`repos/${repository}/git/refs/heads/${branch}`, { sha: commit.sha, force: false }, 'PATCH');
    console.log(`Refreshed ${existing[0].html_url}; ordinary PR CI must pass for ${commit.sha}`);
    return;
  }
  request(`repos/${repository}/git/refs`, { ref: `refs/heads/${branch}`, sha: commit.sha });
  const pr = request(`repos/${repository}/pulls`, { title: 'Update validated Jankurai component revisions', head: branch, base: 'main',
    body: 'Automated family update. Changed components have successful required checks and immutable CI tags. Candidate locks passed combined integration in a disposable CI checkout. Ordinary PR CI must pass before a protected merge.' });
  console.log(pr.html_url);
}
function merge(family) {
  const actor = api('user').login, hub = family.repos.find(repo => repo.name === 'jankurai');
  for (const pr of api(`repos/${repository}/pulls?state=open&base=main&per_page=100`)) {
    if (pr.user.login !== actor || pr.head.repo?.full_name !== repository || !pr.head.ref.startsWith('automation/family-') || pr.draft) continue;
    const changed = api(`repos/${repository}/pulls/${pr.number}/files?per_page=100`);
    if (!changed.length || changed.some(file => !files.includes(file.filename) || file.status !== 'modified')) continue;
    const candidate = Object.fromEntries(files.map(file => [file, content(file, pr.head.sha)]));
    const base = Object.fromEntries(files.map(file => [file, content(file, pr.base.sha)]));
    if (branchFor(candidate) !== pr.head.ref) continue;
    validateCandidate(family, candidate, base);
    if (!successful(hub, pr.head.sha)) continue;
    const result = api(`repos/${repository}/pulls/${pr.number}/merge`, { sha: pr.head.sha, merge_method: 'squash' }, 'PUT');
    if (!result.merged) throw new Error('protected updater merge refused');
    console.log(`Merged ${pr.html_url} at ${result.sha}`);
  }
}
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    const family = new Family(path.dirname(path.dirname(fileURLToPath(import.meta.url))));
    if (process.argv[2] === 'publish') publish(family, path.join(family.hub, 'target/candidate'));
    else if (process.argv[2] === 'merge') merge(family);
    else throw new Error('expected publish or merge');
  } catch (error) { console.error(`family automation: ${error.message}`); process.exitCode = 1; }
}
